sources/Application/Instruments/SampleInstrument.cpp

```diff
@@ -645,6 +647,14 @@ bool SampleInstrument::Render(int channel, fixed *buffer, int size,
               fpSpeed = rp->speed_;
             }
             break;
+          case SILM_LOOP_PINGPONG:
+             if (rp->position_ >= rp->rendLoopEnd_) {
+             // if (loopPosition >= lastSample) {
+               rpReverse = !rpReverse;
+               fpSpeed = -fpSpeed;
+               rp->couldClick_=SHOULD_KILL_CLICKS ;
```

Comment:
Haven't reviewed the full code for SampleInstrument, the one thing that I see that might be a problem is the following:
Shouldn't "rp->position_" be adjusted here to either be the same as rp->rendLoopEnd_ or the offset assuming that it already reverted? (I.e: rp->position_ = rp->rendLoopEnd_ - (rp->position_ - rp->rendLoopEnd_) )

Assuming this works because it's being adjusted somewhere else if rp->position_ > rp->rendLoopEnd_

Reply:
Thanks for the tip!
I'll try it out, don't think I've tried that particular operation.
I know for sure I've tried rp->position_ = rp->rendLoopEnd_ but i've not tried the offset operation.

It looks a little strange to me though, wouldn't this put position at rp->rendLoopEnd_*2 - rp->position_ ?

Reply:
Tried your suggestions on a PSP build, can't see that it changes anything unfortunately.
The issue I still see is that when looping backwards (i.e when the end of the loop is less than the start of the loop) it behaves as a regular loop.
Another thing I've observed is that at the ends of a properly working pingpong, I hear a small glitch of something as if it's reading outside the sample area. :thinking:

Reply:
Does clicking happen in all cases you tried?